datalad/interface/utils.py

```diff
@@ -127,83 +132,312 @@ def handle_dirty_datasets(dpaths,
         raise ValueError("unknown if-dirty mode '{}'".format(mode))
 
 
+def get_tree_roots(paths):
+    """Return common root paths for a set of paths
+
+    This function determines the smallest set of common root
+    paths and sorts all given paths under the respective
+    root.
+
+    Returns
+    -------
+    dict
+      paths by root
+    """
+    paths_ws = [_with_sep(p) for p in paths]
+    # sort all paths under their potential roots
+    roots = {}
+    # start from the top to get all paths down the line
+    # and collate them into as few roots as possible
+    for s in sorted(paths_ws):
+        if any([s.startswith(r) for r in roots]):
+            # this path is already covered by a known root
+            continue
+        # find all sub paths
+        subs = [p for p in paths if p.startswith(s)]
```

Comment:
Yes, that's the point! If we found something starts with `s_1` we already know it's not starting with `s_2`. ;-)